In [1]:
from collections import defaultdict, OrderedDict
from csv import reader
from itertools import chain, combinations
from optparse import OptionParser
import pandas as pd

# if not installed yet: pip install mlxtend
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import fpgrowth, association_rules

###  How to find frequent itemsets using FPGrowth?

In [2]:
dataset = [['Corn', 'Light Cream', 'Chicken', 'Beef', 'Wine', 'Ice Cream'],
           ['Dill', 'Onion', 'Carrot', 'Beef', 'Wine', 'Ice Cream'],
           ['Milk', 'Wine', 'Beef', 'Ice Cream'],
           ['Light Cream', 'Chicken', 'Corn', 'Kidney Beans', 'Yogurt', 'Wine'],
           ['Corn', 'Onion', 'Light Cream', 'Kidney Beans', 'Chicken', 'Yogurt']]

te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)

frequent_itemsets = fpgrowth(df, min_support=0.6, use_colnames=True)
frequent_itemsets

support                      itemsets
0       0.8                        (Wine)
1       0.6                 (Light Cream)
2       0.6                   (Ice Cream)
3       0.6                        (Corn)
4       0.6                     (Chicken)
5       0.6                        (Beef)
6       0.6             (Ice Cream, Wine)
7       0.6           (Corn, Light Cream)
8       0.6               (Corn, Chicken)
9       0.6        (Light Cream, Chicken)
10      0.6  (Corn, Light Cream, Chicken)
11      0.6             (Ice Cream, Beef)
12      0.6                  (Beef, Wine)
13      0.6       (Ice Cream, Beef, Wine)

## Part 1(a): 
### What association rules can be found in given dataset with the minimum support 60% and the minimum confidence 70% ?
If we are only interested in rules with the confidence level above the 70 percent threshold (min_threshold=0.7), we can find the set of rules with the following specifications accordingly using *generate_rules()* function.
1. Specify your metric of interest 
2. Threshold

In [3]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)
rules


antecedents             consequents  antecedent support  \
0              (Ice Cream)                  (Wine)                 0.6   
1                   (Wine)             (Ice Cream)                 0.8   
2                   (Corn)           (Light Cream)                 0.6   
3            (Light Cream)                  (Corn)                 0.6   
4                   (Corn)               (Chicken)                 0.6   
5                (Chicken)                  (Corn)                 0.6   
6            (Light Cream)               (Chicken)                 0.6   
7                (Chicken)           (Light Cream)                 0.6   
8      (Corn, Light Cream)               (Chicken)                 0.6   
9          (Corn, Chicken)           (Light Cream)                 0.6   
10  (Light Cream, Chicken)                  (Corn)                 0.6   
11                  (Corn)  (Light Cream, Chicken)                 0.6   
12           (Light Cream)         (Corn, Chicken)                 0.6   
13               (Chicken)     (Corn, Light Cream)                 0.6   
14             (Ice Cream)                  (Beef)                 0.6   
15                  (Beef)             (Ice Cream)                 0.6   
16                  (Beef)                  (Wine)                 0.6   
17                  (Wine)                  (Beef)                 0.8   
18       (Ice Cream, Beef)                  (Wine)                 0.6   
19       (Ice Cream, Wine)                  (Beef)                 0.6   
20            (Beef, Wine)             (Ice Cream)                 0.6   
21             (Ice Cream)            (Beef, Wine)                 0.6   
22                  (Beef)       (Ice Cream, Wine)                 0.6   
23                  (Wine)       (Ice Cream, Beef)                 0.8   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.8      0.6        1.00  1.250000      0.12         inf  
1                  0.6      0.6        0.75  1.250000      0.12         1.6  
2                  0.6      0.6        1.00  1.666667      0.24         inf  
3                  0.6      0.6        1.00  1.666667      0.24         inf  
4                  0.6      0.6        1.00  1.666667      0.24         inf  
5                  0.6      0.6        1.00  1.666667      0.24         inf  
6                  0.6      0.6        1.00  1.666667      0.24         inf  
7                  0.6      0.6        1.00  1.666667      0.24         inf  
8                  0.6      0.6        1.00  1.666667      0.24         inf  
9                  0.6      0.6        1.00  1.666667      0.24         inf  
10                 0.6      0.6        1.00  1.666667      0.24         inf  
11                 0.6      0.6        1.00  1.666667      0.24         inf  
12                 0.6      0.6        1.00  1.666667      0.24         inf  
13                 0.6      0.6        1.00  1.666667      0.24         inf  
14                 0.6      0.6        1.00  1.666667      0.24         inf  
15                 0.6      0.6        1.00  1.666667      0.24         inf  
16                 0.8      0.6        1.00  1.250000      0.12         inf  
17                 0.6      0.6        0.75  1.250000      0.12         1.6  
18                 0.8      0.6        1.00  1.250000      0.12         inf  
19                 0.6      0.6        1.00  1.666667      0.24         inf  
20                 0.6      0.6        1.00  1.666667      0.24         inf  
21                 0.6      0.6        1.00  1.666667      0.24         inf  
22                 0.6      0.6        1.00  1.666667      0.24         inf  
23                 0.6      0.6        0.75  1.250000      0.12         1.6

## Part 1(b): 
### What association rules can be found in given dataset with the minimum support 60% and the minimum lift value 1.2 ?

In [4]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.2)
rules

antecedents             consequents  antecedent support  \
0              (Ice Cream)                  (Wine)                 0.6   
1                   (Wine)             (Ice Cream)                 0.8   
2                   (Corn)           (Light Cream)                 0.6   
3            (Light Cream)                  (Corn)                 0.6   
4                   (Corn)               (Chicken)                 0.6   
5                (Chicken)                  (Corn)                 0.6   
6            (Light Cream)               (Chicken)                 0.6   
7                (Chicken)           (Light Cream)                 0.6   
8      (Corn, Light Cream)               (Chicken)                 0.6   
9          (Corn, Chicken)           (Light Cream)                 0.6   
10  (Light Cream, Chicken)                  (Corn)                 0.6   
11                  (Corn)  (Light Cream, Chicken)                 0.6   
12           (Light Cream)         (Corn, Chicken)                 0.6   
13               (Chicken)     (Corn, Light Cream)                 0.6   
14             (Ice Cream)                  (Beef)                 0.6   
15                  (Beef)             (Ice Cream)                 0.6   
16                  (Beef)                  (Wine)                 0.6   
17                  (Wine)                  (Beef)                 0.8   
18       (Ice Cream, Beef)                  (Wine)                 0.6   
19       (Ice Cream, Wine)                  (Beef)                 0.6   
20            (Beef, Wine)             (Ice Cream)                 0.6   
21             (Ice Cream)            (Beef, Wine)                 0.6   
22                  (Beef)       (Ice Cream, Wine)                 0.6   
23                  (Wine)       (Ice Cream, Beef)                 0.8   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.8      0.6        1.00  1.250000      0.12         inf  
1                  0.6      0.6        0.75  1.250000      0.12         1.6  
2                  0.6      0.6        1.00  1.666667      0.24         inf  
3                  0.6      0.6        1.00  1.666667      0.24         inf  
4                  0.6      0.6        1.00  1.666667      0.24         inf  
5                  0.6      0.6        1.00  1.666667      0.24         inf  
6                  0.6      0.6        1.00  1.666667      0.24         inf  
7                  0.6      0.6        1.00  1.666667      0.24         inf  
8                  0.6      0.6        1.00  1.666667      0.24         inf  
9                  0.6      0.6        1.00  1.666667      0.24         inf  
10                 0.6      0.6        1.00  1.666667      0.24         inf  
11                 0.6      0.6        1.00  1.666667      0.24         inf  
12                 0.6      0.6        1.00  1.666667      0.24         inf  
13                 0.6      0.6        1.00  1.666667      0.24         inf  
14                 0.6      0.6        1.00  1.666667      0.24         inf  
15                 0.6      0.6        1.00  1.666667      0.24         inf  
16                 0.8      0.6        1.00  1.250000      0.12         inf  
17                 0.6      0.6        0.75  1.250000      0.12         1.6  
18                 0.8      0.6        1.00  1.250000      0.12         inf  
19                 0.6      0.6        1.00  1.666667      0.24         inf  
20                 0.6      0.6        1.00  1.666667      0.24         inf  
21                 0.6      0.6        1.00  1.666667      0.24         inf  
22                 0.6      0.6        1.00  1.666667      0.24         inf  
23                 0.6      0.6        0.75  1.250000      0.12         1.6

## Part 1(c):
### Examine the differences from the two sets of associations rules above. Select one rule that has low confidence and high lift value. Why is this rule interesting?

### Answer:
The rule I choose is (Wine) as antecedents and (beef) as consequents.   
• support of (Wine) is 0.8  
• suport of (beef) is 0.6  
• confidence = support of rule/ support(Wine) = 0.6/0.8 = 0.75  
• lift = confidence / support(beef) = 0.75/(0.6)= 1.25  
Reason: The people who bought beef is few, and takes up 60% of all the subjects. Although only 75% of people who bought Wine bought beef, the low suport of beef will increase the lift. Support of beef is 0.6 while 0.75 of people who bought wine, bought beef, therefore the rule is interesting.

## Part 1(d):
### Describing results: A summary of results (number of rules, general description), and a selection of rules that you deem is interesting based on “interestingness" measure(s) of your choice. Describe your reasoning behind any choices you made.

### Answer:
The selection of interesting rules is based on minimun support of 0.6 and minimun lift is 1.5. And I got 17 rules. Each rule has confidence equal to 1. Thus all rules are interesting. I will interprete one interesting rule: 100% of people who bought Corn also bought Light Cream.  
The reason of this selection is that the lift is greater than 1 to make sure the result is more comfirmed. Also raising the lift can reduce the output.

In [5]:
frequent_itemsets = fpgrowth(df, min_support=0.6,use_colnames=True)
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.5)
print(len(rules["confidence"] >=1))
rules

18


antecedents             consequents  antecedent support  \
0                   (Corn)           (Light Cream)                 0.6   
1            (Light Cream)                  (Corn)                 0.6   
2                   (Corn)               (Chicken)                 0.6   
3                (Chicken)                  (Corn)                 0.6   
4            (Light Cream)               (Chicken)                 0.6   
5                (Chicken)           (Light Cream)                 0.6   
6      (Corn, Light Cream)               (Chicken)                 0.6   
7          (Corn, Chicken)           (Light Cream)                 0.6   
8   (Light Cream, Chicken)                  (Corn)                 0.6   
9                   (Corn)  (Light Cream, Chicken)                 0.6   
10           (Light Cream)         (Corn, Chicken)                 0.6   
11               (Chicken)     (Corn, Light Cream)                 0.6   
12             (Ice Cream)                  (Beef)                 0.6   
13                  (Beef)             (Ice Cream)                 0.6   
14       (Ice Cream, Wine)                  (Beef)                 0.6   
15            (Beef, Wine)             (Ice Cream)                 0.6   
16             (Ice Cream)            (Beef, Wine)                 0.6   
17                  (Beef)       (Ice Cream, Wine)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.6         1.0  1.666667      0.24         inf  
1                  0.6      0.6         1.0  1.666667      0.24         inf  
2                  0.6      0.6         1.0  1.666667      0.24         inf  
3                  0.6      0.6         1.0  1.666667      0.24         inf  
4                  0.6      0.6         1.0  1.666667      0.24         inf  
5                  0.6      0.6         1.0  1.666667      0.24         inf  
6                  0.6      0.6         1.0  1.666667      0.24         inf  
7                  0.6      0.6         1.0  1.666667      0.24         inf  
8                  0.6      0.6         1.0  1.666667      0.24         inf  
9                  0.6      0.6         1.0  1.666667      0.24         inf  
10                 0.6      0.6         1.0  1.666667      0.24         inf  
11                 0.6      0.6         1.0  1.666667      0.24         inf  
12                 0.6      0.6         1.0  1.666667      0.24         inf  
13                 0.6      0.6         1.0  1.666667      0.24         inf  
14                 0.6      0.6         1.0  1.666667      0.24         inf  
15                 0.6      0.6         1.0  1.666667      0.24         inf  
16                 0.6      0.6         1.0  1.666667      0.24         inf  
17                 0.6      0.6         1.0  1.666667      0.24         inf

# Part 2:

In [6]:
class Node:
    def __init__(self, itemName, frequency, parentNode):
        self.itemName = itemName
        self.count = frequency
        self.parent = parentNode
        self.children = {}
        self.next = None

    def increment(self, frequency):
        self.count += frequency

    def display(self, ind=1):
        print('  ' * ind, self.itemName, ' ', self.count)
        for child in list(self.children.values()):
            child.display(ind+1)


def constructTree(itemSetList, frequency, minSup):
    headerTable = defaultdict(int)
    # Counting frequency and create header table
    for idx, itemSet in enumerate(itemSetList):
        for item in itemSet:
            headerTable[item] += frequency[idx]

    # Deleting items below minSup
    headerTable = dict((item, sup) for item, sup in headerTable.items() if sup >= minSup)
    if(len(headerTable) == 0):
        return None, None

    # HeaderTable column [Item: [frequency, headNode]]
    for item in headerTable:
        headerTable[item] = [headerTable[item], None]

    # Init Null head node
    fpTree = Node('Null', 1, None)
    # Update FP tree for each cleaned and sorted itemSet
    for idx, itemSet in enumerate(itemSetList):
        itemSet = [item for item in itemSet if item in headerTable]
################fixed sort function#################
        itemSet.sort(key=lambda item: (headerTable[item][0], item), reverse=True)       
################fixed sort function end############# 
        # Traverse from root to leaf, update tree with given item
        currentNode = fpTree
        for item in itemSet:
            currentNode = updateTree(item, currentNode, headerTable, frequency[idx])

    return fpTree, headerTable

def updateHeaderTable(item, targetNode, headerTable):
    if(headerTable[item][1] == None):
        headerTable[item][1] = targetNode
    else:
        currentNode = headerTable[item][1]
        # Traverse to the last node then link it to the target
        while currentNode.next != None:
            currentNode = currentNode.next
        currentNode.next = targetNode

def updateTree(item, treeNode, headerTable, frequency):
    if item in treeNode.children:
        # If the item already exists, increment the count
        treeNode.children[item].increment(frequency)
    else:
        # Create a new branch
        newItemNode = Node(item, frequency, treeNode)
        treeNode.children[item] = newItemNode
        # Link the new branch to header table
        updateHeaderTable(item, newItemNode, headerTable)

    return treeNode.children[item]

def ascendFPtree(node, prefixPath):
    if node.parent != None:
        prefixPath.append(node.itemName)
        ascendFPtree(node.parent, prefixPath)

def findPrefixPath(basePat, headerTable):
    # First node in linked list
    treeNode = headerTable[basePat][1] 
    condPats = []
    frequency = []
    while treeNode != None:
        prefixPath = []
        # From leaf node all the way to root
        ascendFPtree(treeNode, prefixPath)  
        if len(prefixPath) > 1:
            # Storing the prefix path and it's corresponding count
            condPats.append(prefixPath[1:])
            frequency.append(treeNode.count)

        # Go to next node
        treeNode = treeNode.next  
    return condPats, frequency

def mineTree(headerTable, minSup, preFix, freqItemList):
    # Sort the items with frequency and create a list
    sortedItemList = [item[0] for item in sorted(list(headerTable.items()), key=lambda p:p[1][0])] 
    # Start with the lowest frequency
    for item in sortedItemList:  
        # Pattern growth is achieved by the concatenation of suffix pattern with frequent patterns generated from conditional FP-tree
        newFreqSet = preFix.copy()
        newFreqSet.add(item)
        freqItemList.append(newFreqSet)
        # Find all prefix path, constrcut conditional pattern base
        conditionalPattBase, frequency = findPrefixPath(item, headerTable) 
        # Construct conditonal FP Tree with conditional pattern base
        conditionalTree, newHeaderTable = constructTree(conditionalPattBase, frequency, minSup) 
        if newHeaderTable != None:
            # Mining recursively on the tree
            mineTree(newHeaderTable, minSup,
                       newFreqSet, freqItemList)

def powerset(s):
    return chain.from_iterable(combinations(s, r) for r in range(1, len(s)))

def getSupport(testSet, itemSetList):
    count = 0
    for itemSet in itemSetList:
        if(set(testSet).issubset(itemSet)):
            count += 1
    return count

def associationRule(freqItemSet, itemSetList, minConf):
    rules = []
    for itemSet in freqItemSet:
        subsets = powerset(itemSet)
        itemSetSup = getSupport(itemSet, itemSetList)
        for s in subsets:
            confidence = float(itemSetSup / getSupport(s, itemSetList))
            if(confidence > minConf):
                rules.append([set(s), set(itemSet.difference(s)), confidence])
    return rules

def getFrequencyFromList(itemSetList):
    frequency = [1 for i in range(len(itemSetList))]
    return frequency

In [7]:
def fpgrowth(itemSetList, minSupRatio, minConf):
    frequency = getFrequencyFromList(itemSetList)
    minSup = len(itemSetList) * minSupRatio
    fpTree, headerTable = constructTree(itemSetList, frequency, minSup)
    
    if(fpTree == None):
        print('No frequent item set')
    else:
        freqItems = []
        mineTree(headerTable, minSup, set(), freqItems)
        rules = associationRule(freqItems, itemSetList, minConf)
        return freqItems, rules

## Part 2(a):
### What association rules can be found with the code provided in Part 2 with the minimum support 60% and the minimum confidence 70%? Comparing to the output from question 1(a), what do you think the cause is that leads to different outcome?

### Answer:
The rules are:
``` 
 [[{'Chicken'}, {'Light Cream'}, 1.0],  
  [{'Light Cream'}, {'Chicken'}, 1.0],  
  [{'Wine'}, {'Beef'}, 0.75],  
  [{'Beef'}, {'Wine'}, 1.0],  
  [{'Beef'}, {'Ice Cream'}, 1.0],  
  [{'Ice Cream'}, {'Beef'}, 1.0],  
  [{'Wine'}, {'Ice Cream'}, 0.75],  
  [{'Ice Cream'}, {'Wine'}, 1.0],  
  [{'Wine'}, {'Beef', 'Ice Cream'}, 0.75],  
  [{'Beef'}, {'Ice Cream', 'Wine'}, 1.0],  
  [{'Ice Cream'}, {'Beef', 'Wine'}, 1.0],  
  [{'Beef', 'Wine'}, {'Ice Cream'}, 1.0],  
  [{'Ice Cream', 'Wine'}, {'Beef'}, 1.0],  
  [{'Beef', 'Ice Cream'}, {'Wine'}, 1.0]])
```
Compared to the output from question 1(a), the number of rules are different, and the pf-tree generated is different.  
**Reason**: the sort result is different while constructing the tree. The order of each person's shopping list is only based on the frequency of the item in the whole list. If different item have the same frequency, the item will not be sorted. And the order of node inserted is the order of the shopping list, which is random. That is why when selecting the nodes, some node will be deleted by mistake. Because the frequency of the node is bellow the threshold due to the unsorted names.

In [8]:
freqItemSet, rules = fpgrowth(dataset, minSupRatio=0.6, minConf=0.7)
freqItemSet, rules

([{'Corn'},
  {'Corn', 'Light Cream'},
  {'Light Cream'},
  {'Chicken'},
  {'Chicken', 'Corn'},
  {'Chicken', 'Corn', 'Light Cream'},
  {'Chicken', 'Light Cream'},
  {'Beef'},
  {'Beef', 'Ice Cream'},
  {'Beef', 'Ice Cream', 'Wine'},
  {'Beef', 'Wine'},
  {'Ice Cream'},
  {'Ice Cream', 'Wine'},
  {'Wine'}],
 [[{'Corn'}, {'Light Cream'}, 1.0],
  [{'Light Cream'}, {'Corn'}, 1.0],
  [{'Corn'}, {'Chicken'}, 1.0],
  [{'Chicken'}, {'Corn'}, 1.0],
  [{'Corn'}, {'Chicken', 'Light Cream'}, 1.0],
  [{'Light Cream'}, {'Chicken', 'Corn'}, 1.0],
  [{'Chicken'}, {'Corn', 'Light Cream'}, 1.0],
  [{'Corn', 'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken', 'Corn'}, {'Light Cream'}, 1.0],
  [{'Chicken', 'Light Cream'}, {'Corn'}, 1.0],
  [{'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken'}, {'Light Cream'}, 1.0],
  [{'Ice Cream'}, {'Beef'}, 1.0],
  [{'Beef'}, {'Ice Cream'}, 1.0],
  [{'Ice Cream'}, {'Beef', 'Wine'}, 1.0],
  [{'Beef'}, {'Ice Cream', 'Wine'}, 1.0],
  [{'Wine'}, {'Beef', 'Ice Cream'}, 0.75],


## Part 2(b):
### Fix the problem in the Part 2 code and re-run your program on the given dataset to obtain the same results as in Part 1(a). What did you change?

### Answer:
I changed the key of sort function in the `constructTree`. The mistake was it only sort the list based on the frequency. 
The code was:  
`itemSet.sort(key=lambda item: headerTable[item][0], reverse=True)`  

I add the another parameter to the lambda function. First, sort the table based on the frequency, then sort the name if the value have the same frequency. The fixed sort statement is:  
` itemSet.sort(key=lambda item: (headerTable[item][0], item), reverse=True) `

In [9]:
freqItemSet, rules = fpgrowth(dataset, minSupRatio=0.6, minConf=0.7)
freqItemSet, rules

([{'Corn'},
  {'Corn', 'Light Cream'},
  {'Light Cream'},
  {'Chicken'},
  {'Chicken', 'Corn'},
  {'Chicken', 'Corn', 'Light Cream'},
  {'Chicken', 'Light Cream'},
  {'Beef'},
  {'Beef', 'Ice Cream'},
  {'Beef', 'Ice Cream', 'Wine'},
  {'Beef', 'Wine'},
  {'Ice Cream'},
  {'Ice Cream', 'Wine'},
  {'Wine'}],
 [[{'Corn'}, {'Light Cream'}, 1.0],
  [{'Light Cream'}, {'Corn'}, 1.0],
  [{'Corn'}, {'Chicken'}, 1.0],
  [{'Chicken'}, {'Corn'}, 1.0],
  [{'Corn'}, {'Chicken', 'Light Cream'}, 1.0],
  [{'Light Cream'}, {'Chicken', 'Corn'}, 1.0],
  [{'Chicken'}, {'Corn', 'Light Cream'}, 1.0],
  [{'Corn', 'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken', 'Corn'}, {'Light Cream'}, 1.0],
  [{'Chicken', 'Light Cream'}, {'Corn'}, 1.0],
  [{'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken'}, {'Light Cream'}, 1.0],
  [{'Ice Cream'}, {'Beef'}, 1.0],
  [{'Beef'}, {'Ice Cream'}, 1.0],
  [{'Ice Cream'}, {'Beef', 'Wine'}, 1.0],
  [{'Beef'}, {'Ice Cream', 'Wine'}, 1.0],
  [{'Wine'}, {'Beef', 'Ice Cream'}, 0.75],


## Part 2(c):
### Explain in your own words how the anti-monotonicity property of itemsets is maintained to reduce the search space in generating frequent itemsets.

### Answer:
Anti-monotonicity property According to anti-monotonicity property, if an item is not frequent, all the itemsets containing it are not frequent. Then we can stop further mining for the itemset. In other words, if constraint c is not valid, we can prevent further mining.
In FP-Growth, when the first time the dataset is scanned, if an item appears fewer times than the minimum threshold, it will be removed. Thus, we reduce the frequent itemsets.  
In the condition pattern base step, we will delete the node that is not frequent and reduce the searching space.

## Part 2(d):
### Explain in your own words how the pattern growth is achieved in generating frequent itemsets.

### Answer:
The pattern growth is achieved by recursively grow frequent pattern. First, it start mining from the length-1 patterns, then recursively repeat the process that first construct the condition pattern-tree, then concatenate patterns from conditional FP-tree with suffix. The recursion stops until the subtree is empty or only one path left. Then generate frequent itemsets.

## Part 2(e):
### Explain in your own words how many data scan is required for FPgrowth compared to Apriori.

### Answer:
For FP-growth model, first time we scan the dataset, and find all the length-1 frequent items and the number of each in the dataset. Then remove the ones whose frequency is less than threshold, and sort the rest by descending order.  
Then, second time we scan the data, we remove the non-frequent length-1 items, and create the fp-tree and header table.  
Apriori need to scan the whole dataset every time it find a frequent support.

## Part 2(f):
### Explain in your own words how time complexity and memory consumption is reduced compared to Apriori.

### Answer:
Time complexity: Apriori needs many times to scan the database, which is high IO cost. While FP-growth only need twice.
Memory complexity: Apriori will generate generates a huge candidate set. And the number of candidate sets is exponential growing. FP-growth compress the items in a tree, which reduces a lot of memory.

# Extra Marks
### Please provide a different item ordering (i.e., not descending order of frequency) in constructing FP-Tree and draw the corresponding FP-Tree ofthe same datasetas above. Do you believe that you have the most compressed (or compact) FP-Tree representation in this particular case? 

### Answer:
The order of the data is bellow:
```
dataset = [['Wine', 'Beef', 'Ice Cream', 'corn', 'Light Cream', 'Chicken'],
           ['Wine', 'Beef',  'Ice Cream'],
           ['Wine', 'Beef',  'Ice Cream'],
           [ 'Wine', 'Corn', 'Light Cream', 'Chicken', ],
           ['Corn', 'Light Cream', 'Chicken']]
```
And follow the list, the tree can be contructed as:
```
           null
          /      \   
    Wine 4        Corn 1
   /      \          \  
Beef 3    Corn 1   Light creame 1  
|             \             \  
Ice cream 3   Ice Cream 1      Chicken 1  
|               \
Corn 2          Chicken 1
|  
Light cream 2  
|  
Chicken 2  
```